### PR TITLE
ENH: Modify user attributes with rules

### PR DESCRIPTION
**Pull request by jaidevd**
Use the following spec:

```yaml
url:
    auth/db:
    pattern: /$YAMLURL/db
    handler: DBAuth
    kwargs:
      url: $YAMLPATH/users.csv
      user:
        column: user
      password:
        column: password
      redirect:
        query: next
        header: Referer
        url: .
      rules:
        url: $YAMLPATH/rules.csv
```

And the following two files:

1. [users.csv](https://github.com/gramener/gramex/files/7299389/users.csv)

2. [`rules.csv`](https://github.com/gramener/gramex/files/7299388/rules.csv)

When logging in, each matching rule will be applied to the current user.

### CODE REVIEW

File: gramex/handlers/authhandler.py [hunk 1]
```diff
@@ -70,6 +73,10 @@ def setup(cls, prepare=None, action=None, delay=None, session_expiry=None,
             else:
                 app_log.error('%s: lookup must be a dict, not %s', cls.name, cls.lookup)
 
+        if rules:
+            rules = gramex.data.filter(**rules)
+        cls.rules = rules
```
Comment (jaidevd) [inline gramex/handlers/authhandler.py:78]: If rules is None, create an empty dataframe.

File: tests/test_auth.py [hunk 2]
```diff
@@ -483,6 +483,62 @@ def test_lookup(self):
         eq_(session['user']['gender'], 'female')
 
 
+class TestRules(AuthBase):
+
+    @classmethod
+    def setUpClass(cls):
```
Comment (sanand0) [inline tests/test_auth.py:489]: Another approach is to just set url = server.base_url + '/auth/rulesdb' as a class variable and not override setupClass at all. Same for TestRulesFile

File: tests/gramex.yaml [hunk 3]
```diff
@@ -1993,6 +1993,70 @@ url:
...
+          password: beta
+          gender: male
+          email: john.doe@gramener.com
+          team: "ग्रामेक्स"
+        gamma:
+          password: gamma
+          gender: male
+          empty: ''
+      rules:
+        url: $YAMLPATH/rules.xlsx
```
Comment (sanand0) [inline tests/gramex.yaml]: Please make this a rules.csv -- that way, at least one file is inspectable

File: tests/gramex.yaml [hunk 4]
```diff
@@ -1993,6 +1993,70 @@ url:
...
+        alpha:
+          password: alpha
+          gender: female
+          email: jane.doe@gmail.com
+        beta:
+          password: beta
+          gender: male
+          email: john.doe@gramener.com
+          team: "ग्रामेक्स"
+        gamma:
```
Comment (sanand0) [inline tests/gramex.yaml:2032]: Could this username contain unicode instead of gamma?

File: tests/test_auth.py [hunk 5]
```diff
@@ -483,6 +483,62 @@ def test_lookup(self):
...
+        eq_(session['user']['gender'], 'female')
+        eq_(session['user']['team'], 'Gramex')
+
+        # When selector is a nonexistent attribute, nothing changes
+        self.login_ok('gamma', 'gamma', check_next='/')
+        session = self.session.get(server.base_url + '/auth/session').json()
+        eq_(session['user']['gender'], 'female')
+
+        # Check if the empty string match for Gamma works
+        eq_(session['user']['email'], 'gamma@null.com')
```
Comment (sanand0) [inline tests/test_auth.py:509]: Add test case where we can add field="expiry" with a value= of date type. Also a field="cookie_expires" with value= of float type

Comment (sanand0) [reply tests/test_auth.py:509]: Please also test that no user has an email none@none.com nor gamma@null.com.

Also test that team=Gramex for all
